pkg/reconciler/tenancy/workspacemounts/workspacemounts_controller_workspaces.go

```diff
@@ -0,0 +1,310 @@
+	})
+
+	_, _ = workspaceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueue(obj) },
+		UpdateFunc: func(_, obj interface{}) { c.enqueue(obj) },
+	})
+
+	_, _ = globalShardInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    func(obj interface{}) { c.enqueueShard(obj) },
+		UpdateFunc: func(obj, _ interface{}) { c.enqueueShard(obj) },
+		DeleteFunc: func(obj interface{}) { c.enqueueShard(obj) },
+	})
+
+	c.discoveringDynamicSharedInformerFactory.AddEventHandler(informer.GVREventHandlerFuncs{
+		AddFunc:    func(gvr schema.GroupVersionResource, obj interface{}) { c.enqueueForResource(logger, gvr, obj) },
+		UpdateFunc: func(gvr schema.GroupVersionResource, _, obj interface{}) { c.enqueueForResource(logger, gvr, obj) },
+		DeleteFunc: nil, // Nothing to do.
+	})
+
+	return c, nil
+}
+
+type workspaceResource = committer.Resource[*tenancyv1alpha1.WorkspaceSpec, *tenancyv1alpha1.WorkspaceStatus]
+
+// Controller watches Workspaces and WorkspaceShards in order to make sure every Workspace
+// is scheduled to a valid Shard.
+type Controller struct {
+	// queue is the work-queue used by the controller for workspaces
+	queue workqueue.RateLimitingInterface
+	// gvrQueue is the work-queue used by the controller for dynamic resources (mounts)
+	gvrQueue workqueue.RateLimitingInterface
+
+	shardName                         string
+	logicalClusterAdminConfig         *rest.Config // for direct shard connections used during scheduling
+	externalLogicalClusterAdminConfig *rest.Config // for front-proxy connections used during initialization
+
+	kcpClusterClient     kcpclientset.ClusterInterface
+	kubeClusterClient    kubernetes.ClusterInterface
+	kcpExternalClient    kcpclientset.ClusterInterface
+	dynamicClusterClient kcpdynamic.ClusterInterface
+
+	workspaceIndexer cache.Indexer
+	workspaceLister  tenancyv1alpha1listers.WorkspaceClusterLister
+
+	globalShardIndexer cache.Indexer
+	globalShardLister  corev1alpha1listers.ShardClusterLister
+
+	globalWorkspaceTypeIndexer cache.Indexer
+	globalWorkspaceTypeLister  tenancyv1alpha1listers.WorkspaceTypeClusterLister
+
+	logicalClusterIndexer cache.Indexer
+	logicalClusterLister  corev1alpha1listers.LogicalClusterClusterLister
+
+	discoveringDynamicSharedInformerFactory *informer.DiscoveringDynamicSharedInformerFactory
+
+	// commit creates a patch and submits it, if needed.
+	commit func(ctx context.Context, new, old *workspaceResource) error
+}
+
+// enqueue adds the object to the work queue, used for workspaces.
+func (c *Controller) enqueue(obj interface{}) {
+	key, err := kcpcache.MetaClusterNamespaceKeyFunc(obj)
+	if err != nil {
+		runtime.HandleError(err)
+		return
+	}
+	logger := logging.WithQueueKey(logging.WithReconciler(klog.Background(), ControllerName), key)
+	logger.V(2).Info("queueing Workspace")
+	c.queue.Add(key)
+}
+
+func (c *Controller) enqueueShard(obj interface{}) {
```

Comment:
why enqueing shards?

Reply:
originally was thinking we might need to do this to trigger all worksapce in the shard. Now I learned we dont need to do it :) Had wrong mental model in my head